File: lib_common/src/main/java/com/wujia/lib_common/data/network/DownLoadManager.java
package com.wujia.lib_common.data.network;

import android.os.Handler;
import android.os.Looper;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import okhttp3.ResponseBody;

public class DownLoadManager {

    private DownloadCallBack callBack;

    private static final String TAG = "DownLoadManager";

    private Handler handler;

    public DownLoadManager(DownloadCallBack callBack) {
        this.callBack = callBack;
    }

    private static DownLoadManager sInstance;

    /**
     * DownLoadManager getInstance
     */
    public static synchronized DownLoadManager getInstance(DownloadCallBack callBack) {
        if (sInstance == null) {
            sInstance = new DownLoadManager(callBack);
        }
        return sInstance;
    }


    public boolean writeResponseBodyToDisk(String filePath, ResponseBody body) {


        final String path = filePath;


        try {
            File futureStudioIconFile = new File(path);

            if (futureStudioIconFile.exists()) {
                futureStudioIconFile.delete();
            }

            InputStream inputStream = null;
            OutputStream outputStream = null;

            try {
                byte[] fileReader = new byte[4096];

                final long fileSize = body.contentLength();
                long fileSizeDownloaded = 0;
                inputStream = body.byteStream();
                outputStream = new FileOutputStream(futureStudioIconFile);

                while (true) {
                    int read = inputStream.read(fileReader);

                    if (read == -1) {
                        break;
                    }

                    outputStream.write(fileReader, 0, read);

                    fileSizeDownloaded += read;

                    if (callBack != null) {
                        handler = new Handler(Looper.getMainLooper());
                        final long finalFileSizeDownloaded = fileSizeDownloaded;
                        handler.post(new Runnable() {
                            @Override
                            public void run() {
                                callBack.onProgress((int) (100 * finalFileSizeDownloaded / fileSize));
                            }
                        });

                    }
                }

                outputStream.flush();
                if (callBack != null) {
                    handler = new Handler(Looper.getMainLooper());
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            callBack.onSucess(path, fileSize);
                        }
                    });
                }

                return true;
            } catch (IOException e) {
                if (callBack != null) {
                    callBack.onError(e);
                }
                return false;
            } finally {
                if (inputStream != null) {
                    inputStream.close();
                }

                if (outputStream != null) {
                    outputStream.close();
                }
            }
        } catch (IOException e) {
            if (callBack != null) {
                callBack.onError(e);
            }
            return false;
        }
    }
}

File: lib_common/src/main/java/com/wujia/lib_common/data/network/DownLoadManager.kt
package com.wujia.lib_common.data.network

import android.os.Handler
import android.os.Looper
import okhttp3.ResponseBody
import java.io.*

class DownLoadManager(private val callBack: DownloadCallBack?) {

    private var handler: Handler? = null


    fun writeResponseBodyToDisk(filePath: String, body: ResponseBody): Boolean {


        try {
            val futureStudioIconFile = File(filePath)

            if (futureStudioIconFile.exists()) {
                futureStudioIconFile.delete()
            }

            var inputStream: InputStream? = null
            var outputStream: OutputStream? = null

            try {
                val fileReader = ByteArray(4096)

                val fileSize = body.contentLength()
                var fileSizeDownloaded: Long = 0
                inputStream = body.byteStream()
                outputStream = FileOutputStream(futureStudioIconFile)

                while (true) {
                    val read = inputStream!!.read(fileReader)

                    if (read == -1) {
                        break
                    }

                    outputStream.write(fileReader, 0, read)

                    fileSizeDownloaded += read.toLong()

                    if (callBack != null) {
                        handler = Handler(Looper.getMainLooper())
                        val finalFileSizeDownloaded = fileSizeDownloaded
                        handler!!.post { callBack.onProgress((100 * finalFileSizeDownloaded / fileSize).toInt().toLong()) }

                    }
                }

                outputStream.flush()
                if (callBack != null) {
                    handler = Handler(Looper.getMainLooper())
                    handler!!.post { callBack.onSucess(filePath, fileSize) }
                }

                return true
            } catch (e: IOException) {
                callBack?.onError(e)
                return false
            } finally {
                inputStream?.close()

                outputStream?.close()
            }
        } catch (e: IOException) {
            callBack?.onError(e)
            return false
        }

    }

    companion object {

        private val TAG = "DownLoadManager"

        private var sInstance: DownLoadManager? = null

        /**
         * DownLoadManager getInstance
         */
        @Synchronized
        fun getInstance(callBack: DownloadCallBack?): DownLoadManager {
            if (sInstance == null) {
                sInstance = DownLoadManager(callBack)
            }
            return sInstance!!
        }
    }
}
